include/xtensor/xtiny.hpp

```diff
@@ -0,0 +1,1999 @@
+               V2 const & r)
+    {
+        for(decltype(l.size()) k=0; k < l.size(); ++k)
+            if(l[k] != r)
+                return false;
+        return true;
+    }
+
+    template <class V1, class V2, index_t N2, class R2,
+              XTENSOR_REQUIRE<!xtiny_concept<V1>::value &&
+                              std::is_convertible<V2, V1>::value> >
+    inline bool
+    operator==(V1 const & l,
+               xtiny<V2, N2, R2> const & r)
+    {
+        for(decltype(r.size()) k=0; k < r.size(); ++k)
+            if(l != r[k])
+                return false;
+        return true;
+    }
+
+    template <class V1, index_t N1, class R1, class V2, index_t N2, class R2>
+    inline bool
+    operator!=(xtiny<V1, N1, R1> const & l,
+               xtiny<V2, N2, R2> const & r)
+    {
+        if(l.size() != r.size())
+            return true;
+        for(decltype(l.size()) k=0; k < l.size(); ++k)
+            if(l[k] != r[k])
+                return true;
+        return false;
+    }
+
+    template <class V1, index_t N1, class R1, class V2,
+              XTENSOR_REQUIRE<!xtiny_concept<V2>::value &&
+                              std::is_convertible<V2, V1>::value> >
+    inline bool
+    operator!=(xtiny<V1, N1, R1> const & l,
+               V2 const & r)
+    {
+        for(decltype(l.size()) k=0; k < l.size(); ++k)
+            if(l[k] != r)
+                return true;
+        return false;
+    }
+
+    template <class V1, class V2, index_t N2, class R2,
+              XTENSOR_REQUIRE<!xtiny_concept<V1>::value &&
+                              std::is_convertible<V2, V1>::value> >
+    inline bool
+    operator!=(V1 const & l,
+               xtiny<V2, N2, R2> const & r)
+    {
+        for(decltype(r.size()) k=0; k < r.size(); ++k)
+            if(l != r[k])
+                return true;
+        return false;
+    }
+
+    /************************/
+    /* xtiny implementation */
+    /************************/
+
+    template <class V, index_t N, class R>
+    inline
+    xtiny<V, N, R>::xtiny()
+    : base_type()
+    {
+    }
+
+    template <class V, index_t N, class R>
+    inline
+    xtiny<V, N, R>::xtiny(xtiny const & v)
+    : base_type(v)
+    {
+    }
+
+    template <class V, index_t N, class R>
+    inline
+    xtiny<V, N, R>::xtiny(xtiny && v)
+    : base_type(std::forward<base_type>(v))
+    {
+    }
+
+    template <class V, index_t N, class R>
+    template <class T, index_t M, class Q>
+    inline
+    xtiny<V, N, R>::xtiny(xtiny<T, M, Q> const & v)
+    : base_type(v.begin(), v.end())
+    {
+    }
+
+    template <class V, index_t N, class R>
+    template <class T, class A>
+    inline
+    xtiny<V, N, R>::xtiny(std::vector<T, A> const & v)
+    : base_type(v.begin(), v.end())
+    {
+    }
+
+    template <class V, index_t N, class R>
+    template <class T, std::size_t M>
+    inline
+    xtiny<V, N, R>::xtiny(std::array<T, M> const & v)
+    : base_type(v.begin(), v.end())
```

Comment:
maybe use cbegin / cend everywhere?

Reply:
Where's the difference?

Reply:
It is more expressive. We use `cbegin`, `cend` wherever possible across the code base. When we find an instance where it is not the case, we change it.

Reply:
changed